ambry-network/src/main/java/com/github/ambry/network/AdaptiveLifoCoDelNetworkRequestQueue.java

```diff
@@ -0,0 +1,169 @@
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * Adaptive LIFO blocking queue utilizing CoDel algorithm to prevent queue overloading. This queue is based on
+ * (and borrows most of its code from) {@code AdaptiveLifoCoDelCallQueue} in HBase.
+ *
+ * @see <a href="http://queue.acm.org/detail.cfm?id=2839461">Fail at Scale paper</a>
+ * @see <a href="https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/AdaptiveLifoCoDelCallQueue.java">HBase implementation</a>
+ */
+public class AdaptiveLifoCoDelNetworkRequestQueue implements NetworkRequestQueue {
+  private final Time time;
+  private final LinkedBlockingDeque<NetworkRequest> deque;
+
+  // so we can calculate actual threshold to switch to LIFO under load
+  private final int maxCapacity;
+  // if queue if full more than that percent, we switch to LIFO mode.
+  // Values are in the range of 0.7, 0.8 etc (0-1.0).
+  private final double lifoThreshold;
+  // Both are in milliseconds
+  private final int coDelTargetDelayMs;
+  private final int coDelIntervalMs;
+  // minimal delay observed during the interval
+  private volatile long minDelayMs;
+  // the moment when current interval ends
+  private volatile long intervalTimeMs;
+  // switch to ensure only one thread does interval cutoffs
+  private final AtomicBoolean resetDelay = new AtomicBoolean(true);
+  // if we're in this mode, "long" calls are getting dropped
+  private final AtomicBoolean isOverloaded = new AtomicBoolean(false);
+
+  /**
+   *
+   * @param capacity the max capacity of the queue.
+   * @param lifoThreshold the fraction of capacity used at which to switch the queue from FIFO to LIFO mode.
+   * @param coDelTargetDelayMs the target delay in ms to use for the controlled delay algorithm.
+   * @param coDelIntervalMs the target interval in ms to use for the controlled delay algorithm.
+   * @param time {@link Time} instance.
+   */
+  AdaptiveLifoCoDelNetworkRequestQueue(int capacity, double lifoThreshold, int coDelTargetDelayMs, int coDelIntervalMs,
+      Time time) {
+    this.maxCapacity = capacity;
+    this.coDelTargetDelayMs = coDelTargetDelayMs;
+    this.coDelIntervalMs = coDelIntervalMs;
+    this.lifoThreshold = lifoThreshold;
+    this.time = time;
+
+    deque = new LinkedBlockingDeque<>(capacity);
+    intervalTimeMs = time.milliseconds();
+  }
+
+  /**
+   * Behaves as {@link LinkedBlockingQueue#take()}, except it will silently
+   * skip all calls which it thinks should be dropped.
+   *
+   * @return the head of this queue
+   * @throws InterruptedException if interrupted while waiting
+   */
+  @Override
+  public NetworkRequestBundle take() throws InterruptedException {
+    NetworkRequest requestToServe = null;
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    while (true) {
+      NetworkRequest nextRequest;
+      if (useLifoMode()) {
+        nextRequest = deque.pollLast();
```

Comment:
Sure. Added log statement to print when mode changes in `useLifoMode` method. For metrics, actually added a set of metrics at higher level to track dropped request rate for various types of requests. Since changing of queue mode is an implementation detail, wondering if metric may not be needed here. What do you say?